lib/tesla/adapter/mint.ex

```diff
@@ -0,0 +1,190 @@
+  end
+
+  defp request(method, scheme, host, port, path, headers, body, opts) when is_function(body) do
+    with {:ok, conn} <- HTTP.connect(String.to_atom(scheme), host, port, opts),
+         # FIXME Stream function in Mint will not append the content length after eof
+         # This will trigger the failure in unit test
+         {:ok, body, length} <- stream_request(body),
+         {:ok, conn, _req_ref} <-
+           HTTP.request(
+             conn,
+             method,
+             path || "/",
+             headers ++ [{"content-length", "#{length}"}],
+             body
+           ),
+         {:ok, conn, res = %{status: status, headers: headers}} <- stream_response(conn, opts),
+         {:ok, _conn} <- HTTP.close(conn) do
+      {:ok, status, headers, Map.get(res, :data)}
+    end
+  end
+
+  defp request(method, scheme, host, port, path, headers, body, opts) do
+    with {:ok, conn} <- HTTP.connect(String.to_atom(scheme), host, port, opts),
+         {:ok, conn, _req_ref} <- HTTP.request(conn, method, path || "/", headers, body),
+         {:ok, conn, res = %{status: status, headers: headers}} <- stream_response(conn, opts),
+         {:ok, _conn} <- HTTP.close(conn) do
+      {:ok, status, headers, Map.get(res, :data)}
+    end
+  end
+
+  defp get_global_default_ca() do
+    env = Application.get_env(:tesla, Tesla.Adapter.Mint)
+    Keyword.get(env, :cacert)
+  end
+
+  defp stream_request(fun, body \\ "") do
+    case next_chunk(fun) do
+      {:ok, item, fun} when is_list(item) ->
+        stream_request(fun, body <> List.to_string(item))
+
+      {:ok, item, fun} ->
+        stream_request(fun, body <> item)
+
+      :eof ->
+        {:ok, body, byte_size(body)}
+    end
+  end
+
+  defp stream_response(conn, opts, response \\ %{}) do
+    receive do
+      msg ->
+        case HTTP.stream(conn, msg) do
+          {:ok, conn, stream} ->
+            response =
+              Enum.reduce(stream, response, fn x, acc ->
+                case x do
+                  {:status, _req_ref, code} ->
+                    Map.put(acc, :status, code)
+
+                  {:headers, _req_ref, headers} ->
+                    Map.put(acc, :headers, Map.get(acc, :headers, []) ++ headers)
+
+                  {:data, _req_ref, data} ->
+                    Map.put(acc, :data, Map.get(acc, :data, "") <> data)
+
+                  {:done, _req_ref} ->
+                    Map.put(acc, :done, true)
+
+                  {:error, _req_ref, reason} ->
+                    Map.put(acc, :error, reason)
+
+                  _ ->
+                    acc
+                end
+              end)
+
+            cond do
+              Map.has_key?(response, :error) ->
+                {:error, Map.get(response, :error)}
+
+              Map.has_key?(response, :done) ->
+                {:ok, conn, Map.drop(response, [:done])}
+
+              true ->
+                stream_response(conn, response, opts)
```

Comment:
Order is not correct.
```suggestion
                stream_response(conn, opts, response)
```